File: pkg/js_grim/lib/installer.js
#!/usr/bin/env node

const https = require('https');
const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const { createReadStream, createWriteStream } = require('fs');
const { pipeline } = require('stream');
const { promisify } = require('util');
const zlib = require('zlib');
const tar = require('tar');

const streamPipeline = promisify(pipeline);

class GrimInstaller {
    constructor() {
        this.installDir = path.join(process.env.HOME || '/root', '.grim-reaper');
        this.downloadUrl = 'https://get.grim.so/latest.tar.gz';
        this.tempFile = path.join(process.env.HOME || '/tmp', 'grim-latest.tar.gz');
    }

    /**
     * Check if Grim is already installed
     */
    isInstalled() {
        const possiblePaths = [
            path.join(this.installDir, 'throne', 'grim_throne.sh'),
            path.join(this.installDir, 'sh_grim', 'backup.sh'),
            path.join(this.installDir, 'py_grim', 'grim_web', 'server.py'),
            path.join(this.installDir, 'go_grim', 'build', 'grim-compression')
        ];

        return possiblePaths.some(p => fs.existsSync(p));
    }

    /**
     * Download Grim tarball from get.grim.so
     */
    async download() {
        console.log('📥 Downloading Grim Reaper from get.grim.so...');
        
        return new Promise((resolve, reject) => {
            const file = fs.createWriteStream(this.tempFile);
            
            https.get(this.downloadUrl, (response) => {
                if (response.statusCode === 302 || response.statusCode === 301) {
                    // Follow redirect
                    https.get(response.headers.location, (redirectResponse) => {
                        redirectResponse.pipe(file);
                        file.on('finish', () => {
                            file.close();
                            console.log('✅ Download complete');
                            resolve();
                        });
                    }).on('error', reject);
                } else if (response.statusCode === 200) {
                    response.pipe(file);
                    file.on('finish', () => {
                        file.close();
                        console.log('✅ Download complete');
                        resolve();
                    });
                } else {
                    reject(new Error(`Failed to download: ${response.statusCode}`));
                }
            }).on('error', reject);
        });
    }

    /**
     * Extract the downloaded tarball
     */
    async extract() {
        console.log('📦 Extracting Grim Reaper...');
        
        // Create install directory
        if (!fs.existsSync(this.installDir)) {
            fs.mkdirSync(this.installDir, { recursive: true });
        }

        // Extract tarball
        await tar.x({
            file: this.tempFile,
            cwd: this.installDir,
            strip: 1 // Remove top-level directory from tarball
        });

        console.log('✅ Extraction complete');
    }

    /**
     * Make scripts executable
     */
    makeExecutable() {
        console.log('🔧 Making scripts executable...');
        
        const scripts = [
            'throne/grim_throne.sh',
            'throne/sh_grim_throne.sh',
            'throne/py_grim_throne.sh',
            'throne/go_grim_throne.sh',
            'sh_grim/*.sh',
            'go_grim/build/*',
            'install.sh',
            'master-install.sh'
        ];

        scripts.forEach(pattern => {
            const fullPattern = path.join(this.installDir, pattern);
            try {
                if (pattern.includes('*')) {
                    // Handle glob patterns
                    const dir = path.dirname(fullPattern);
                    const filePattern = path.basename(fullPattern);
                    
                    if (fs.existsSync(dir)) {
                        fs.readdirSync(dir).forEach(file => {
                            if (filePattern === '*.sh' && file.endsWith('.sh')) {
                                const filePath = path.join(dir, file);
                                fs.chmodSync(filePath, 0o755);
                            } else if (filePattern === '*') {
                                const filePath = path.join(dir, file);
                                if (fs.statSync(filePath).isFile()) {
                                    fs.chmodSync(filePath, 0o755);
                                }
                            }
                        });
                    }
                } else {
                    // Handle specific files
                    if (fs.existsSync(fullPattern)) {
                        fs.chmodSync(fullPattern, 0o755);
                    }
                }
            } catch (error) {
                // Ignore permission errors
            }
        });

        console.log('✅ Scripts made executable');
    }

    /**
     * Install system dependencies
     */
    async installDependencies() {
        console.log('📦 Checking system dependencies...');
        
        const deps = ['rsync', 'tar', 'gzip', 'python3', 'pip3'];
        const missing = [];

        deps.forEach(dep => {
            try {
                execSync(`which ${dep}`, { stdio: 'ignore' });
            } catch {
                missing.push(dep);
            }
        });

        if (missing.length > 0) {
            console.log(`⚠️  Missing dependencies: ${missing.join(', ')}`);
            console.log('💡 Please install them using your package manager');
        }

        console.log('✅ Dependency check complete');
    }

    /**
     * Setup .scythe directory structure
     */ 
    setupScytheDirectories() {
        console.log('🗡️  Setting up .scythe directory structure...');
        
        const scytheDir = path.join(this.installDir, '.graveyard', '.rip', '.scythe');
        const directories = ['config', 'db', 'logs', 'run', 'integrations'];
        
        // Create directory structure
        directories.forEach(dir => {
            const dirPath = path.join(scytheDir, dir);
            if (!fs.existsSync(dirPath)) {
                fs.mkdirSync(dirPath, { recursive: true });
            }
        });
        
        // Create scythe configuration file
        const configFile = path.join(scytheDir, 'config', 'scythe.yaml');
        if (!fs.existsSync(configFile)) {
            const configContent = `# Scythe Configuration
# Central orchestrator settings for Grim Reaper System

scythe:
  version: "1.0.5"
  install_date: ${new Date().toISOString()}
  
database:
  path: "../db/scythe.db"
  auto_backup: true
  backup_interval: "24h"
  
logging:
  level: "info"
  path: "../logs"
  max_size: "100MB"
  max_files: 10
  
orchestration:
  enabled: true
  heartbeat_interval: "30s"
  max_concurrent_jobs: 5
  
integrations:
  enabled: true
  scan_interval: "5m"
  auto_discover: true
  
security:
  encryption: true
  key_rotation: "30d"
  audit_logs: true
`;
            fs.writeFileSync(configFile, configContent);
            console.log('✅ Created scythe configuration');
        }
        
        // Initialize scythe database using shell script if available
        const setupScript = path.join(this.installDir, 'scripts', 'setup_scythe_dirs.sh');
        if (fs.existsSync(setupScript)) {
            try {
                execSync(`bash "${setupScript}" setup "${this.installDir}" no`, { stdio: 'ignore' });
                console.log('✅ Initialized scythe database');
            } catch (error) {
                console.log('⚠️  Could not initialize scythe database - will create basic structure');
                // Create basic directories as fallback
                fs.mkdirSync(path.join(scytheDir, 'logs', 'orchestration'), { recursive: true });
                fs.mkdirSync(path.join(scytheDir, 'logs', 'components'), { recursive: true });
                fs.mkdirSync(path.join(scytheDir, 'logs', 'integrations'), { recursive: true });
                fs.mkdirSync(path.join(scytheDir, 'logs', 'security'), { recursive: true });
                fs.mkdirSync(path.join(scytheDir, 'integrations', 'discovered'), { recursive: true });
                fs.mkdirSync(path.join(scytheDir, 'integrations', 'configs'), { recursive: true });
                fs.mkdirSync(path.join(scytheDir, 'integrations', 'scripts'), { recursive: true });
            }
        }
        
        console.log('✅ .scythe directory structure created');
    }

    /**
     * Setup environment variables
     */
    setupEnvironment() {
        console.log('🌍 Setting up environment...');
        
        const scytheDir = path.join(this.installDir, '.graveyard', '.rip', '.scythe');
        const envContent = `# Grim Reaper Environment
export GRIM_ROOT="${this.installDir}"
export SCYTHE_DIR="${scytheDir}"
export PATH="$GRIM_ROOT/throne:$PATH"
`;

        // Add to shell profile
        const shellProfiles = [
            path.join(process.env.HOME, '.bashrc'),
            path.join(process.env.HOME, '.zshrc'),
            path.join(process.env.HOME, '.profile')
        ];

        shellProfiles.forEach(profile => {
            if (fs.existsSync(profile)) {
                const content = fs.readFileSync(profile, 'utf8');
                if (!content.includes('GRIM_ROOT')) {
                    fs.appendFileSync(profile, '\n' + envContent);
                    console.log(`✅ Updated ${profile}`);
                }
            }
        });

        // Set for current session
        process.env.GRIM_ROOT = this.installDir;
        process.env.SCYTHE_DIR = scytheDir;
        console.log('✅ Environment setup complete');
    }

    /**
     * Run post-install script if available
     */
    async runPostInstall() {
        const postInstallScript = path.join(this.installDir, 'install.sh');
        
        if (fs.existsSync(postInstallScript)) {
            console.log('🚀 Running post-install script...');
            try {
                execSync(`cd "${this.installDir}" && bash install.sh`, { 
                    stdio: 'inherit',
                    env: { ...process.env, GRIM_ROOT: this.installDir }
                });
                console.log('✅ Post-install complete');
            } catch (error) {
                console.log('⚠️  Post-install script failed, but Grim should still work');
            }
        }
    }

    /**
     * Cleanup temporary files
     */
    cleanup() {
        if (fs.existsSync(this.tempFile)) {
            fs.unlinkSync(this.tempFile);
        }
    }

    /**
     * Main installation process
     */
    async install() {
        console.log('🗡️  Grim Reaper Installer');
        console.log('========================\n');

        try {
            // Check if already installed
            if (this.isInstalled()) {
                console.log('✅ Grim Reaper is already installed at:', this.installDir);
                return this.installDir;
            }

            // Download
            await this.download();

            // Extract
            await this.extract();

            // Make executable
            this.makeExecutable();

            // Install dependencies
            await this.installDependencies();

            // Setup .scythe directories
            this.setupScytheDirectories();

            // Setup environment
            this.setupEnvironment();

            // Run post-install
            await this.runPostInstall();

            // Cleanup
            this.cleanup();

            console.log('\n✅ Grim Reaper installation complete!');
            console.log(`📁 Installed to: ${this.installDir}`);
            console.log('💡 Restart your shell or run: source ~/.bashrc');
            console.log('🗡️  Then use: grim <command>\n');

            return this.installDir;

        } catch (error) {
            console.error('❌ Installation failed:', error.message);
            this.cleanup();
            throw error;
        }
    }

    /**
     * Get install directory
     */
    getInstallDir() {
        return this.installDir;
    }
}

module.exports = GrimInstaller;

// Run installer if called directly
if (require.main === module) {
    const installer = new GrimInstaller();
    installer.install().catch(console.error);
}

File: pkg/rb_grim/lib/grim_reaper/installer.rb
# frozen_string_literal: true

require "open3"
require "fileutils"

module GrimReaper
  # Ruby-specific installer and dependency manager
  class Installer
    attr_reader :grim_root, :backup_dir, :install_dir

    def initialize
      @grim_root = find_grim_root
      @backup_dir = ENV["HOME"] + "/.graveyard"
      @install_dir = ENV["HOME"] + "/reaper"
    end

    # Find Grim Reaper root directory
    def find_grim_root
      # Try to find from current directory
      current_dir = Dir.pwd
      10.times do
        if File.exist?(File.join(current_dir, "throne", "grim_throne.sh")) ||
           File.exist?(File.join(current_dir, "throne", "rb_grim_throne.sh")) ||
           File.exist?(File.join(current_dir, "tsk_flask", "grim_admin_server.py"))
          return current_dir
        end

        parent_dir = File.dirname(current_dir)
        break if parent_dir == current_dir
        current_dir = parent_dir
      end

      # Try common installation paths
      possible_paths = [
        ENV["HOME"] + "/reaper",
        ENV["HOME"] + "/.reaper",
        "/root/reaper",
        "/root/.reaper",
        "/usr/local/reaper",
        "/usr/share/reaper",
        Dir.pwd
      ]

      possible_paths.each do |path|
        if Dir.exist?(path) && (
          File.exist?(File.join(path, "throne", "grim_throne.sh")) ||
          File.exist?(File.join(path, "throne", "rb_grim_throne.sh")) ||
          File.exist?(File.join(path, "tsk_flask", "grim_admin_server.py"))
        )
          return path
        end
      end

      raise Error, "Could not find Grim Reaper root directory"
    end

    # Setup complete Grim Reaper environment (like npm/pip packages)
    def setup_complete_environment
      puts "🗡️  Setting up complete Grim Reaper environment..."
      
      # Setup Ruby environment
      setup_ruby_environment
      
      # Install Python components
      install_python_components
      
      # Install Go components
      install_go_components
      
      # Install Shell components
      install_shell_components
      
      # Install Scythe components
      install_scythe_components
      
      # Create necessary directories
      create_directories
      
      # Setup environment file
      setup_environment_file
      
      puts "✅ Complete Grim Reaper environment setup finished"
    end

    # Setup Ruby environment
    def setup_ruby_environment
      puts "💎 Setting up Ruby environment..."

      # Check Ruby version
      check_ruby_version

      # Install Bundler if not present
      install_bundler

      # Install common Ruby gems
      install_ruby_gems

      # Setup RVM/rbenv if available
      setup_ruby_version_manager

      puts "✅ Ruby environment setup complete"
    end

    # Install Python components (like pip package)
    def install_python_components
      puts "🐍 Installing Python components..."
      
      # Check if Python is available
      unless system("python3 --version > /dev/null 2>&1")
        puts "⚠️  Python3 not found, skipping Python components"
        return
      end
      
      # Install pip if not available
      unless system("pip3 --version > /dev/null 2>&1")
        puts "📦 Installing pip3..."
        unless system("apt-get update && apt-get install -y python3-pip") || 
               system("yum install -y python3-pip")
          puts "⚠️  Could not install pip3"
        end
      end
      
      # Install Python dependencies
      python_packages = %w[flask requests psutil cryptography]
      python_packages.each do |package|
        unless system("pip3 list | grep -q #{package}")
          puts "📦 Installing Python package: #{package}"
          unless system("pip3 install #{package}")
            puts "⚠️  Failed to install #{package}"
          end
        end
      end
      
      # Create Python virtual environment if needed
      venv_path = File.join(@grim_root, "py_grim", "venv")
      unless Dir.exist?(venv_path)
        puts "🐍 Creating Python virtual environment..."
        unless system("python3 -m venv #{venv_path}")
          puts "⚠️  Failed to create virtual environment"
        end
      end
      
      puts "✅ Python components setup complete"
    end

    # Install Go components
    def install_go_components
      puts "🐹 Installing Go components..."
      
      # Check if Go is available
      unless system("go version > /dev/null 2>&1")
        puts "📦 Installing Go..."
        unless system("apt-get update && apt-get install -y golang-go") ||
               system("yum install -y golang")
          puts "⚠️  Could not install Go"
        end
        return
      end
      
      # Create Go build directory
      go_build_dir = File.join(@grim_root, "go_grim", "build")
      FileUtils.mkdir_p(go_build_dir) unless Dir.exist?(go_build_dir)
      
      # Build Go binaries if source exists
      go_src_dir = File.join(@grim_root, "go_grim")
      if Dir.exist?(go_src_dir) && File.exist?(File.join(go_src_dir, "go.mod"))
        puts "🐹 Building Go binaries..."
        Dir.chdir(go_src_dir) do
          unless system("go mod tidy")
            puts "⚠️  Failed to tidy Go modules"
          end
          unless system("go build -o build/grim-compression .")
            puts "⚠️  Failed to build Go binary"
          end
        end
      end
      
      puts "✅ Go components setup complete"
    end

    # Install Shell components
    def install_shell_components
      puts "🐚 Installing Shell components..."
      
      # Make shell scripts executable
      shell_scripts = [
        "sh_grim/backup.sh",
        "sh_grim/restore.sh", 
        "sh_grim/monitor.sh",
        "sh_grim/scan.sh",
        "sh_grim/security.sh",
        "throne/grim_throne.sh",
        "throne/rb_grim_throne.sh"
      ]
      
      shell_scripts.each do |script|
        script_path = File.join(@grim_root, script)
        if File.exist?(script_path)
          File.chmod(0o755, script_path)
          puts "🔧 Made executable: #{script}"
        end
      end
      
      # Install system dependencies
      system_deps = %w[rsync tar gzip bzip2 xz-utils]
      system_deps.each do |dep|
        unless system("which #{dep} > /dev/null 2>&1")
          puts "📦 Installing system dependency: #{dep}"
          unless system("apt-get update && apt-get install -y #{dep}") ||
                 system("yum install -y #{dep}")
            puts "⚠️  Could not install #{dep}"
          end
        end
      end
      
      puts "✅ Shell components setup complete"
    end

    # Install Scythe components
    def install_scythe_components
      puts "🗡️  Installing Scythe components..."
      
      # Check if scythe directory exists
      scythe_dir = File.join(@grim_root, "scythe")
      unless Dir.exist?(scythe_dir)
        puts "⚠️  Scythe directory not found, creating..."
        FileUtils.mkdir_p(scythe_dir)
      end
      
      # Install Python dependencies for scythe
      if system("python3 --version > /dev/null 2>&1")
        scythe_packages = %w[flask requests cryptography psutil]
        scythe_packages.each do |package|
          unless system("pip3 list | grep -q #{package}")
            puts "📦 Installing Scythe dependency: #{package}"
            unless system("pip3 install #{package}")
              puts "⚠️  Failed to install #{package}"
            end
          end
        end
      end
      
      puts "✅ Scythe components setup complete"
    end

    # Check Ruby version
    def check_ruby_version
      ruby_version = `ruby -e "puts RUBY_VERSION"`.strip
      puts "💎 Ruby Version: #{ruby_version}"

      major, minor = ruby_version.split(".").map(&:to_i)
      if major < 2 || (major == 2 && minor < 7)
        raise Error, "Ruby 2.7+ required, found #{ruby_version}"
      end
    end

    # Install Bundler
    def install_bundler
      unless system("gem list bundler -i > /dev/null 2>&1")
        puts "📦 Installing Bundler..."
        unless system("gem install bundler")
          raise Error, "Failed to install Bundler"
        end
      end
    end

    # Install common Ruby gems
    def install_ruby_gems
      gems = %w[rake rspec rubocop yard reek brakeman bundle-audit]
      
      gems.each do |gem_name|
        unless system("gem list #{gem_name} -i > /dev/null 2>&1")
          puts "📦 Installing #{gem_name}..."
          unless system("gem install #{gem_name}")
            warn "Failed to install #{gem_name}"
          end
        end
      end
    end

    # Setup Ruby version manager
    def setup_ruby_version_manager
      if system("which rvm > /dev/null 2>&1")
        puts "🔄 RVM detected, setting up..."
        system("rvm use --default")
      elsif system("which rbenv > /dev/null 2>&1")
        puts "🔄 rbenv detected, setting up..."
        system("rbenv rehash")
      end
    end

    # Setup .scythe directory structure
    def setup_scythe_directories
      puts "🗡️  Setting up .scythe directory structure..."
      
      scythe_dir = File.join(@grim_root, ".graveyard", ".rip", ".scythe")
      
      # Create main scythe directories
      scythe_subdirs = [
        "config",
        "db", 
        "logs",
        "run",
        "integrations"
      ]
      
      scythe_subdirs.each do |subdir|
        dir_path = File.join(scythe_dir, subdir)
        FileUtils.mkdir_p(dir_path) unless Dir.exist?(dir_path)
      end
      
      # Create log subdirectories
      log_subdirs = ["orchestration", "components", "integrations", "security"]
      log_subdirs.each do |subdir|
        dir_path = File.join(scythe_dir, "logs", subdir)
        FileUtils.mkdir_p(dir_path) unless Dir.exist?(dir_path)
      end
      
      # Create integration subdirectories
      integration_subdirs = ["discovered", "configs", "scripts"]
      integration_subdirs.each do |subdir|
        dir_path = File.join(scythe_dir, "integrations", subdir)
        FileUtils.mkdir_p(dir_path) unless Dir.exist?(dir_path)
      end
      
      # Create scythe configuration file
      config_file = File.join(scythe_dir, "config", "scythe.yaml")
      unless File.exist?(config_file)
        config_content = <<~CONFIG
          # Scythe Configuration
          # Central orchestrator settings for Grim Reaper System

          scythe:
            version: "1.0.5"
            install_date: #{Time.now.iso8601}
            
          database:
            path: "../db/scythe.db"
            auto_backup: true
            backup_interval: "24h"
            
          logging:
            level: "info"
            path: "../logs"
            max_size: "100MB"
            max_files: 10
            
          orchestration:
            enabled: true
            heartbeat_interval: "30s"
            max_concurrent_jobs: 5
            
          integrations:
            enabled: true
            scan_interval: "5m"
            auto_discover: true
            
          security:
            encryption: true
            key_rotation: "30d"
            audit_logs: true
        CONFIG
        
        File.write(config_file, config_content)
        puts "✅ Created scythe configuration: #{config_file}"
      end
      
      # Try to run the universal setup script if available
      setup_script = File.join(@grim_root, "scripts", "setup_scythe_dirs.sh")
      if File.exist?(setup_script)
        begin
          system("bash '#{setup_script}' setup '#{@grim_root}' auto")
          puts "✅ Initialized scythe database"
        rescue
          puts "⚠️  Could not initialize scythe database - basic structure created"
        end
      end
      
      puts "✅ .scythe directory structure created at: #{scythe_dir}"
    end

    # Create necessary directories
    def create_directories
      # Setup .scythe structure first
      setup_scythe_directories
      
      dirs = [
        File.join(@grim_root, "logs"),
        File.join(@grim_root, "cache"),
        @backup_dir,
        File.join(@grim_root, "temp"),
        File.join(@grim_root, "py_grim", "venv"),
        File.join(@grim_root, "go_grim", "build"),
        File.join(@grim_root, "scythe", "logs")
      ]

      dirs.each do |dir|
        FileUtils.mkdir_p(dir) unless Dir.exist?(dir)
        puts "📁 Created directory: #{dir}"
      end
    end

    # Setup environment file
    def setup_environment_file
      env_file = File.join(@grim_root, ".env")
      scythe_dir = File.join(@grim_root, ".graveyard", ".rip", ".scythe")
      env_content = <<~ENV
        # Grim Reaper Environment Configuration
        GRIM_ROOT=#{@grim_root}
        SCYTHE_DIR=#{scythe_dir}
        GRIM_BACKUP_DIR=#{@backup_dir}
        GRIM_HOME=#{@install_dir}
        GRIM_RUBY_VERSION=#{`ruby -e "puts RUBY_VERSION"`.strip}
        GRIM_PYTHON_VERSION=#{`python3 --version 2>/dev/null | cut -d' ' -f2` || 'Not installed'}
        GRIM_GO_VERSION=#{`go version 2>/dev/null | cut -d' ' -f3` || 'Not installed'}
      ENV

      File.write(env_file, env_content)
      puts "📄 Created environment file: #{env_file}"
    end

    # Verify installation
    def verify_installation
      puts "🔍 Verifying installation..."

      checks = [
        { name: "Ruby", check: -> { system("ruby --version > /dev/null 2>&1") } },
        { name: "Bundler", check: -> { system("bundle --version > /dev/null 2>&1") } },
        { name: "Python3", check: -> { system("python3 --version > /dev/null 2>&1") } },
        { name: "Go", check: -> { system("go version > /dev/null 2>&1") } },
        { name: "Grim Root", check: -> { Dir.exist?(@grim_root) } },
        { name: "Ruby Throne", check: -> { File.exist?(File.join(@grim_root, "throne", "rb_grim_throne.sh")) } },
        { name: "Python Throne", check: -> { File.exist?(File.join(@grim_root, "throne", "py_grim_throne.sh")) } },
        { name: "Go Throne", check: -> { File.exist?(File.join(@grim_root, "throne", "go_grim_throne.sh")) } },
        { name: "Shell Throne", check: -> { File.exist?(File.join(@grim_root, "throne", "sh_grim_throne.sh")) } },
        { name: "Backup Directory", check: -> { Dir.exist?(@backup_dir) } }
      ]

      checks.each do |check|
        if check[:check].call
          puts "✅ #{check[:name]}: OK"
        else
          puts "❌ #{check[:name]}: FAILED"
        end
      end
    end

    # Get installation status
    def status
      {
        ruby_version: `ruby -e "puts RUBY_VERSION"`.strip,
        bundler_version: `bundle --version 2>/dev/null`.strip,
        python_version: `python3 --version 2>/dev/null | cut -d' ' -f2` || 'Not installed',
        go_version: `go version 2>/dev/null | cut -d' ' -f3` || 'Not installed',
        grim_root: @grim_root,
        backup_dir: @backup_dir,
        install_dir: @install_dir,
        ruby_throne_exists: File.exist?(File.join(@grim_root, "throne", "rb_grim_throne.sh")),
        python_throne_exists: File.exist?(File.join(@grim_root, "throne", "py_grim_throne.sh")),
        go_throne_exists: File.exist?(File.join(@grim_root, "throne", "go_grim_throne.sh")),
        shell_throne_exists: File.exist?(File.join(@grim_root, "throne", "sh_grim_throne.sh")),
        backup_dir_exists: Dir.exist?(@backup_dir)
      }
    end

    # Cleanup temporary files
    def cleanup
      temp_dir = File.join(@grim_root, "temp")
      if Dir.exist?(temp_dir)
        FileUtils.rm_rf(temp_dir)
        puts "🧹 Cleaned temporary directory"
      end
    end

    # Post-install hook
    def self.post_install
      new.setup_complete_environment
    end

    # Post-update hook
    def self.post_update
      new.verify_installation
    end
  end
end 